src/azure-cli/azure/cli/command_modules/acs/custom.py

```diff
@@ -1802,10 +1815,33 @@ def aks_create(cmd, client, resource_group_name, name, ssh_key_value,  # pylint:
                     namespace='Microsoft.ContainerService', type='managedClusters',
                     name=name
                 )
-                if not _add_role_assignment(cmd.cli_ctx, 'Monitoring Metrics Publisher',
-                                            service_principal_profile.client_id, scope=cluster_resource_id):
-                    logger.warning('Could not create a role assignment for monitoring addon. '
-                                   'Are you an Owner on this subscription?')
+
+                service_principal_msi_id = None
+                # Check if service principal exists, if it does, assign permissions to service principal
+                # Else, provide permissions to MSI
+                if (
+                        hasattr(result, 'service_principal_profile') and
+                        hasattr(result.service_principal_profile, 'client_id')
+                ):
+                    logger.warning('service principal exists, using it')
+                    service_principal_msi_id = result.service_principal_profile.client_id
+                    isServicePrincipal = True
+                elif (
+                        (hasattr(result, 'addon_profiles')) and
+                        ('omsagent' in result.addon_profiles) and
+                        (hasattr(result.addon_profiles['omsagent'], 'identity')) and
+                        (hasattr(result.addon_profiles['omsagent'].identity, 'object_id'))
+                ):
+                    logger.warning('omsagent MSI exists, using it')
+                    service_principal_msi_id = result.addon_profiles['omsagent'].identity.object_id
+                    isServicePrincipal = False
+
+                if service_principal_msi_id is not None:
+                    if not _add_role_assignment(cmd.cli_ctx, 'Monitoring Metrics Publisher',
+                                                service_principal_msi_id, isServicePrincipal,
+                                                scope=cluster_resource_id):
+                        logger.warning('Could not create a role assignment for monitoring addon. '
```

Comment:
I think try/except is better than `if not`

Reply:
@MyronFanQiu - I have retained the same behavior as before and with other places where the role assignments are done within this file. Is it a new guideline to use try/catch instead of if not?

Reply:
The method _add_role_assignment returns a boolean value. I think adding try except would changes the behavior of this method and all the depending methods. I will retain the same behavior as before. Unless, there is an absolute need to change it